src/InstructionSelector.cpp

```diff
@@ -0,0 +1,34 @@
+#include "InstructionSelector.h"
+
+#include "CodeGen_Internal.h"
+#include "IROperator.h"
+
+namespace Halide {
+namespace Internal {
+
+InstructionSelector::InstructionSelector(const Target &t, const CodeGen_LLVM *c)
+    : target(t), codegen(c) {
+}
+
+Expr InstructionSelector::visit(const Div *op) {
+    if (op->type.is_vector() && op->type.is_int_or_uint()) {
+        // Lower division here in order to do pattern-matching on intrinsics.
+        return mutate(lower_int_uint_div(op->a, op->b));
+    }
+    return IRGraphMutator::visit(op);
+}
+
+Expr InstructionSelector::visit(const Mod *op) {
+    if (op->type.is_vector() && op->type.is_int_or_uint()) {
+        // Lower mod here in order to do pattern-matching on intrinsics.
+        return mutate(lower_int_uint_mod(op->a, op->b));
+    }
+    return IRGraphMutator::visit(op);
+}
+
+Expr InstructionSelector::visit(const VectorReduce *op) {
+    return mutate(codegen->split_vector_reduce(op, Expr()));
```

Comment:
We should pull split_vector_reduce out of CodeGen_LLVM into the InstructionSelector and have it emit VectorInstruction nodes with enum values that correspond to the things that LLVM natively supports.

This might mean the instruction selector needs to know the native vector width, and maybe how to upgrade types for arithmetic. This could be pulled out of LLVM into common support code, e.g. in CodeGen_Internal.cpp

Reply:
(because currently we believe that this will cause InstructionSelector to emit call nodes to llvm intrinsics)